src/type/definition.js

```diff
@@ -782,6 +853,16 @@ export type GraphQLObjectTypeConfig<TSource, TContext> = {
   extensionASTNodes?: ?$ReadOnlyArray<ObjectTypeExtensionNode>,
 };
 
+export type GraphQLObjectTypeExactConfig<TSource, TContext> = {|
+  name: string,
+  interfaces: Array<GraphQLInterfaceType>,
```

Comment:
Should this instead be:
```
interfaces: Array<GraphQLInterfaceTypeExactConfig<TSource, TContext>>,
```
I'm not sure which makes sense: we definitely don't want to end up with a Schema where an object points to a different instance of the same interface vs. what the schema itself has. So we need to make sure, anywhere we hydrate a type out of this config, that it ends up pointing to the schema's actual interface type, rather than the interface type used at build time (I see you're already doing this everywhere it's used, but others might not be so prudent). I'd even be tempted to make this just `Array<string>` just to prevent any mismatches. At that point, we're basically be creating a stricter version of the Introspection schema (we seem to need a non-Introspection, but Introspection-like, representation regardless, so I'm still OK with this).

My worry may be a complete non-issue if we don't allow exporting these new types and limit how we use them within the repo.